File: src/test/java/com/swarnim/prs/model/GameTest.java
package com.swarnim.prs.model;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

class GameTest {
    @Test
    void testGameLogic_PlayerWins() {
        MoveStrategy playerStrategy = mock(MoveStrategy.class);
        when(playerStrategy.chooseMove()).thenReturn(Move.PAPER);

        MoveStrategy computerStrategy = mock(MoveStrategy.class);
        when(computerStrategy.chooseMove()).thenReturn(Move.ROCK);

        Game game = new Game(playerStrategy, computerStrategy);
        game.playRound();

        assertEquals(1, game.getPlayerWins());
        assertEquals(0, game.getComputerWins());
        assertEquals(0, game.getTies());
    }

    @Test
    void testGameLogic_ComputerWins() {
        MoveStrategy playerStrategy = mock(MoveStrategy.class);
        when(playerStrategy.chooseMove()).thenReturn(Move.SCISSORS);

        MoveStrategy computerStrategy = mock(MoveStrategy.class);
        when(computerStrategy.chooseMove()).thenReturn(Move.ROCK);

        Game game = new Game(playerStrategy, computerStrategy);
        game.playRound();

        assertEquals(0, game.getPlayerWins());
        assertEquals(1, game.getComputerWins());
        assertEquals(0, game.getTies());
    }

    @Test
    void testGameLogic_Tie() {
        MoveStrategy playerStrategy = mock(MoveStrategy.class);
        when(playerStrategy.chooseMove()).thenReturn(Move.ROCK);

        MoveStrategy computerStrategy = mock(MoveStrategy.class);
        when(computerStrategy.chooseMove()).thenReturn(Move.ROCK);

        Game game = new Game(playerStrategy, computerStrategy);
        game.playRound();

        assertEquals(0, game.getPlayerWins());
        assertEquals(0, game.getComputerWins());
        assertEquals(1, game.getTies());
    }
}

File: src/test/kotlin/com/hartwig/paperrockscissors/game/GameTest.kt
package com.hartwig.paperrockscissors.game

import com.hartwig.paperrockscissors.core.GameResult
import com.hartwig.paperrockscissors.core.Move
import org.junit.jupiter.api.Test
import org.mockito.kotlin.*
import com.hartwig.paperrockscissors.players.Player
import org.junit.jupiter.api.BeforeEach

internal class GameTest {

    private var playerA: Player = mock()
    private var playerB: Player = mock()
    private var game = Game()

    @BeforeEach
    fun reset() {
        reset(playerA)
        reset(playerB)
    }

    @Test
    fun testRockRock() {
        whenever(playerA.move()).thenReturn(Move.ROCK)
        whenever(playerB.move()).thenReturn(Move.ROCK)
        game.play(playerA, playerB)
        verify(playerA).move()
        verify(playerB).move()
        verify(playerA).recordGameResult(eq(GameResult.DRAW))
        verify(playerB).recordGameResult(eq(GameResult.DRAW))
    }

    @Test
    fun testRockScissors() {
        whenever(playerA.move()).thenReturn(Move.ROCK)
        whenever(playerB.move()).thenReturn(Move.SCISSORS)
        game.play(playerA, playerB)
        verify(playerA).move()
        verify(playerB).move()
        verify(playerA).recordGameResult(eq(GameResult.WIN))
        verify(playerB).recordGameResult(eq(GameResult.LOSE))
    }

    @Test
    fun testScissorsRock() {
        whenever(playerA.move()).thenReturn(Move.SCISSORS)
        whenever(playerB.move()).thenReturn(Move.ROCK)
        game.play(playerA, playerB)
        verify(playerA).move()
        verify(playerB).move()
        verify(playerA).recordGameResult(eq(GameResult.LOSE))
        verify(playerB).recordGameResult(eq(GameResult.WIN))
    }
}
